### DISCUSSION TITLE
How does exllama_hf use less VRAM?

### DISCUSSION THREAD

**Discussion by bartowski1182** [oobabooga/text-generation-webui#3248] [General]
As far as I can find, all it does is replace the logits and the samples/pipeline (though even that part confuses me), so why does it result in lower consumption? Anyone know where I can read up on it?

**Comment by matatonic** [comment]
There was a comment in the exllama pull request which went into detail, but essentially from what I understood, the top memory gains were made by not fragmenting memory by not dynamically growing the memory (it allocates max sizes at start). There may be other improvements, but that was the big one afaik.

**Reply by 78Alpha** [reply]
That sounds like memory pinning. A feature used a lot in dream booth training. Also helps with speed.
